### PR TITLE
Attachments are implemented

### PR DESCRIPTION
**Pull request by onuridrisoglu**

### CODE REVIEW

File: bugrap/src/main/java/org/vaadin/bugrap/BugrapUI.java [hunk 1]
```diff
@@ -37,6 +37,8 @@ protected void init(VaadinRequest vaadinRequest) {
 		final ReportsView reportsView = new ReportsView(new ReportsModel(navigator));
 		navigator.addView(BaseModel.NAV_REPORT, reportsView);
 		navigator.navigateTo(BaseModel.NAV_LOGIN);
+		
+		navigator.navigateTo(BaseModel.NAV_LOGIN );
```
Comment (marciopd) [inline bugrap/src/main/java/org/vaadin/bugrap/BugrapUI.java]: Seems that this line is repeated.

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+import com.vaadin.ui.Upload.ProgressListener;
+import com.vaadin.ui.Upload.Receiver;
+import com.vaadin.ui.Upload.StartedEvent;
+import com.vaadin.ui.Upload.StartedListener;
+import com.vaadin.ui.Upload.SucceededEvent;
+import com.vaadin.ui.Upload.SucceededListener;
+import com.vaadin.ui.VerticalLayout;
+import com.vaadin.ui.Window;
+
+public class ReportDetailView extends ReportDetailViewBase implements Receiver, SucceededListener, StartedListener, ProgressListener{
```
Comment (marciopd) [inline bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java]: There are some formatting issues in the file. Consider formatting the whole file.

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+		initializeComboContents();
+		fillComments();
+	}
+	
+	private void initializeComboContents() {
+		cmbEditPriority.setItems(model.getPriorties());
+		cmbEditType.setItems(model.getTypes());
+		cmbEditStatus.setItems(model.getStatuses());
+		cmbEditAssignedTo.setItems(model.findReporters());
+		cmbEditVersion.setItems(model.findProjectVersions(model.getReportForEdit().getProject()));
```
Comment (marciopd) [inline bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java]: The model is already aware of the project.

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+	}
+
+
+	@Override
+	public void uploadStarted(StartedEvent event) {
+		pnlAttachments.setVisible(true);
+		HorizontalLayout layout = (HorizontalLayout) pnlAttachments.getContent();
+		ProgressBar progress = new ProgressBar();
+		progress.setCaption(event.getFilename());
+		model.getAttachmentUIElements().put(event.getFilename(), progress);
```
Comment (marciopd) [inline bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java]: What happens if two attachments have the same filename? Is there something else you could be using as the map key?

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+		HorizontalLayout layout = (HorizontalLayout) pnlAttachments.getContent();
+		layout.removeComponent(progress);
+		
+		FileResource file = new FileResource(new File(ReportsModel.FILEUPLOAD_PATH + event.getFilename()));
+		Link link = new Link(event.getFilename(), file);
+		link.setStyleName("tiny");
+		layout.addComponent(link);
+		model.getAttachmentUIElements().put(event.getFilename(), link);
+		try {
+			model.saveAttachment(event.getFilename(), event.getMIMEType(), file.getStream());
```
Comment (marciopd) [inline bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java]: The save should be done only when the save button is pressed.

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+
+	@Override
+	public void updateProgress(long readBytes, long contentLength) {
+		ProgressBar bar = null;
+		for (Object obj : model.getAttachmentUIElements().values()) {
+			if (obj instanceof ProgressBar)
+				bar = (ProgressBar)obj;
+		}
+		if (bar != null)
```
Comment (marciopd) [inline bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java]: Is always only the last progress bar being updated here?

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ReportsModel.java [hunk 7]
```diff
@@ -52,6 +61,10 @@ else if (selectedReports.size() == 1)
 			return SELECTIONMODE_MULTI;
```
Comment (marciopd) [inline bugrap/src/main/java/org/vaadin/bugrap/ui/ReportsModel.java:63]: Please consider formatting the whole file.

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ReportsView.java [hunk 8]
```diff
@@ -140,4 +156,15 @@ private void revertChanges() {
...
+	
+	private void openReportDetail() {
+		Window window = new Window();
+		window.setCaption(String.format("%s > %s", model.getReportForEdit().getProject().getName(), model.getReportForEdit().getVersion().getVersion()));
+		window.setSizeFull();
+		
+		ReportDetailView view = new ReportDetailView(model);
+		window.setContent(view);
+		
+		getUI().addWindow(window);
```
Comment (marciopd) [inline bugrap/src/main/java/org/vaadin/bugrap/ui/ReportsView.java]: Remember using the specific Vaadin component here.

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ThreadItemView.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+	private void init() {
+		lblAuthor.setCaption(String.format("%s (%s)", model.getAuthorText(), model.getLastUpdateText()));
+		if (model.getComment().getType() == Type.COMMENT) {
+			lblComment.setValue(model.getComment().getComment());
+			lblComment.setVisible(true);
+			lnkAttachment.setVisible(false);
+		} else if (model.getComment().getType() == Type.ATTACHMENT) {
+			lblComment.setVisible(false);
+			lnkAttachment.setVisible(true);
+			lnkAttachment.setCaption(model.getComment().getAttachmentName());
```
Comment (marciopd) [inline bugrap/src/main/java/org/vaadin/bugrap/ui/ThreadItemView.java:30]: There's specific Vaadin 8 component for file downloads. You should try it out here.

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+		btnUpload.addStartedListener(this);
+		btnUpload.addSucceededListener(this);
+		btnUpload.addProgressListener(this);
+	}
+
+	private void init() {
+		((HorizontalLayout)pnlAttachments.getContent()).removeAllComponents();
+		pnlAttachments.setVisible(false);
+		Report report = model.getReportForEdit();
+		binder.setBean(model.getReportForEdit());
```
Comment (marciopd) [inline bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java]: Calling twice the same method here.

**Review by marciopd — CHANGES_REQUESTED** [review]

File: bugrap/src/main/java/org/vaadin/bugrap/BugrapUI.java (hunk 1 above)
Comment (onuridrisoglu) [reply bugrap/src/main/java/org/vaadin/bugrap/BugrapUI.java]: removed

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java (hunk 2 above)
Comment (onuridrisoglu) [reply bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java]: done

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java (hunk 10 above)
Comment (onuridrisoglu) [reply bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java]: done

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java (hunk 3 above)
Comment (onuridrisoglu) [reply bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java]: changed

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ThreadItemView.java (hunk 9 above)
Comment (onuridrisoglu) [reply bugrap/src/main/java/org/vaadin/bugrap/ui/ThreadItemView.java:30]: Changed to FileDownloader

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ReportsView.java (hunk 8 above)
Comment (onuridrisoglu) [reply bugrap/src/main/java/org/vaadin/bugrap/ui/ReportsView.java]: Do you mean BrowserWindowOpener?

**Comment by marciopd** [comment]
Hi @onuridrisoglu , please resolve the conflicts before I review the changes.

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java [hunk 11]
```diff
@@ -55,119 +55,122 @@ private void initializeUIComponents() {
...
 		binder.setBean(model.getReportForEdit());
 	}
 
 	private void saveComment() {
-		if (!txtComment.getOptionalValue().isPresent())
-			return;
-		model.saveComment(txtComment.getValue(), BaseModel.loginUser);
+		if (!txtComment.isEmpty())
+			model.saveComment(txtComment.getValue(), BaseModel.loginUser);
+		model.saveAttachments();
```
Comment (marciopd) [inline bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java:116]: What happens if there is a attachment still being uploaded when the user clicks Save?

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java [hunk 12]
```diff
@@ -55,119 +55,122 @@ private void initializeUIComponents() {
...
 	}
 
-
 	@Override
 	public void updateProgress(long readBytes, long contentLength) {
 		ProgressBar bar = null;
-		for (Object obj : model.getAttachmentUIElements().values()) {
-			if (obj instanceof ProgressBar)
-				bar = (ProgressBar)obj;
+		for (Object obj : model.getUploadingUIElements().values()) {
```
Comment (marciopd) [inline bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java]: I didn't understand what this for is about..

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java (hunk 6 above)
Comment (marciopd) [reply bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java]: No changes here yet. Seems that it continues to update only the last progress bar. What if we have two concurrent uploads?

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ReportsView.java (hunk 8 above)
Comment (marciopd) [reply bugrap/src/main/java/org/vaadin/bugrap/ui/ReportsView.java]: Yes.

**Review by marciopd — CHANGES_REQUESTED** [review]
There are a couple of things still missing, all the rest looks good.

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java (hunk 11 above)
Comment (onuridrisoglu) [reply bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java:116]: done button is disabled when upload starts

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java (hunk 12 above)
Comment (onuridrisoglu) [reply bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java]: changes made for single upload

File: bugrap/src/main/java/org/vaadin/bugrap/ui/ReportsModel.java [hunk 13]
```diff
@@ -36,7 +37,7 @@
 
 	private List<Report> selectedReports = new ArrayList<Report>();
 	private List<Comment> uploadedFilesToSave = new ArrayList<Comment>();
-	private Map<String, Object> uploadingUIElements = new HashMap<String, Object>();
+//	private Map<String, Object> uploadingUIElements = new HashMap<String, Object>();
```
Comment (marciopd) [inline bugrap/src/main/java/org/vaadin/bugrap/ui/ReportsModel.java]: Just as a comment, avoid leaving commented code.

**Review by marciopd — APPROVED** [review]
Looks good!

**Pull request merged**